### ISSUE TITLE
FSErrors freeze Python when thrown in multiprocessing context 

### ISSUE THREAD

**Issue by Spacerat** [PyFilesystem/pyfilesystem2#183]
I am using:
- Python 3.6.4
- fs==2.0.24

I discovered this issue when trying to use `listdir` in the context of a concurrent.futures.ProcessPoolExecutor. When a child of `FSError` is thrown in a multiprocessing context, the program never exits.

Here is a minimal reproduction

```py
import concurrent.futures
import fs.errors

def inner():
    # A simple ValueError would be bubbled to the parent process correctly
    raise fs.errors.ResourceError("foo")

def main():
    with concurrent.futures.ProcessPoolExecutor() as pool:
        pool.submit(inner).result()
        # Program never exits.


if __name__ == "__main__":
    main()
```

The following error is displayed, and the program hangs:
```
Exception in thread Thread-1:
Traceback (most recent call last):
  File "/usr/local/Cellar/python/3.6.4_4/Frameworks/Python.framework/Versions/3.6/lib/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/usr/local/Cellar/python/3.6.4_4/Frameworks/Python.framework/Versions/3.6/lib/python3.6/threading.py", line 864, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/local/Cellar/python/3.6.4_4/Frameworks/Python.framework/Versions/3.6/lib/python3.6/concurrent/futures/process.py", line 272, in _queue_management_worker
    result_item = reader.recv()
  File "/usr/local/Cellar/python/3.6.4_4/Frameworks/Python.framework/Versions/3.6/lib/python3.6/multiprocessing/connection.py", line 251, in recv
    return _ForkingPickler.loads(buf.getbuffer())
TypeError: __init__() missing 1 required positional argument: 'path'
```

I have narrowed the implementations of the pyfilesystem2 errors down to the minimum required to trigger the problem:

```py
class MyFSError(Exception):
    def __init__(self):
        # Calling __init__ as __init__(str(self)) here seems to fix the problem.
        # but it only works if all child exceptions have only a single argument.
        super(MyFSError, self).__init__() 


class MyResourceError(MyFSError):
    def __init__(self, path):
        # Supplying a default argument to MyResourceError also fixes the problem.
        super(MyResourceError, self).__init__()

def inner():
    raise MyResourceError("foo")
    # Note that `raise MyFSError()` does not cause the same problem
```

I don't _fully_ understand how this is breaking the multiprocessing machinery except that it is a pickleability problem.

It appears one workable fix is to supply a default `path` argument for `ResourceError`, however, I'm not sure this solution is 'technically' correct. I googled around and found [this thread](https://bugs.python.org/issue32034) where people who seem to have stumbled across the same problem decided that the default argument solution was wrong.

Their solution was to implement `__reduce__` on the child error, as suggested [here](https://stackoverflow.com/questions/16244923/how-to-make-a-custom-exception-class-with-multiple-init-args-pickleable). It seems to me that it may also be possible to implement it on the parent error.

The following snippet demonstrates all of the possible solutions.

1. Use optional arguments in child exceptions
2. Implement `__reduce__` in all child exceptions
3. My attempt at a `__reduce__` implementation in the parent exception

```py
import concurrent.futures

class FixedFSError(Exception):
    default_message = ""

    def __init__(self, msg=None):
        # self._args = tuple(self.__dict__.values()) # SOLUTON 3
        self._msg = self.default_message
        super(FixedFSError, self).__init__()

    def __str__(self):
        return self._msg.format(**self.__dict__)

    # def __reduce__(self): # SOLUTION 3
    #     return type(self), self._args  # SOLUTION 3


class FixedResourceError(FixedFSError):
    default_message = "Failed to find {path}"

    def __init__(self, path):
    # def __init__(self, path=None): # SOLUTION 1
        self.path = path
        super().__init__()

    # def __reduce__(self):  # SOLUTON 2
    #    return type(self), (self.path,)  # SOLUTON 2


def possible_fixes():
    raise FixedResourceError("foo")


def main():
    with concurrent.futures.ProcessPoolExecutor() as pool:
        pool.submit(possible_fixes).result()


if __name__ == "__main__":
    main()
```

**Comment by willmcgugan** [comment]
Thanks for doing the research.

Solution 2 feels like the way to go to me. Shouldn't be too much effort, since all the exceptions are confined to a single file. I'd accept a PR for this, if you are interested.

**Comment by Spacerat** [comment]
Hi @willmcgugan - I opened a PR for this, #184 . Please let me know if you need anything else in order to get it merged

**Comment by willmcgugan** [comment]
Fixed in 2.0.26
